### PR TITLE
Handle non-absolute cairoPath

### LINKED ISSUE #305: Handle cairoPaths for non-absolute paths
Sample `cairoPaths` could look like this 
```js
cairoPaths: ["./contracts/starknet/starknet-dss", "./contracts/starknet/starknet-dss-teleport"]
```

### PR DESCRIPTION
**Pull request by Nathan-SL**
## Usage related changes

-   Closes #305

## Development related changes

-   NA

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors + tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [x] Rebased to the last commit of the target branch (or merged it into my branch)
-   [ ] Documented the changes
-   [x] Updated the `test` directory (with a test case consisting of `network.json`, `hardhat.config.ts`, `check.ts`)
-   [x] Linked issues which this PR resolves
-   [ ] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/Shard-Labs/starknet-hardhat-example):
    -   < EXAMPLE_REPO_PR_URL > 
    -   [ ] Modified `test.sh` to use my example repo branch
    -   [ ] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [ ] All tests are passing (for external contributors who don't have access to the CI/CD pipeline)

### CODE REVIEW

File: src/task-actions.ts [hunk 1]
```diff
@@ -88,8 +88,11 @@ export async function starknetCompileAction(args: TaskArguments, hre: HardhatRun
         });
     }
     if (hre.config.paths.cairoPaths) {
-        hre.config.paths.cairoPaths.forEach((path: string) => {
-            cairoPaths.push(path);
+        hre.config.paths.cairoPaths.forEach((cairPath: string) => {
```
Comment (FabijanC) [inline src/task-actions.ts:91]: Typo? cairPath?

File: src/task-actions.ts [hunk 2]
```diff
@@ -88,8 +88,11 @@ export async function starknetCompileAction(args: TaskArguments, hre: HardhatRun
...
+            if (!path.isAbsolute(cairPath)) {
```
Comment (FabijanC) [inline src/task-actions.ts:92]: This check and the "fixing" of the path in the next line, should be done for all members of the final `cairoPaths` array. But looking at L98, we already do that. So I'm puzzled why this failed in the first place. Therefore we should introduce a test for this (probably can be a simple one, without the need to extend the example repo). We should first see if the test fails without your fix, and then if it passes with the fix.

File: src/task-actions.ts [hunk 3]
```diff
@@ -88,8 +88,11 @@ export async function starknetCompileAction(args: TaskArguments, hre: HardhatRun
...
+                cairPath = path.join(process.env.PWD, cairPath);
```
Comment (FabijanC) [inline src/task-actions.ts:93]: Should it be `process.env.PWD` or `hre.config.paths.root` as we already do in L100? Also, `path.normalize` could be important.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: src/task-actions.ts (hunk 3 above)
Comment (Nathan-SL) [reply src/task-actions.ts:93]: I think `hre.config.paths.root` is better for consistency's sake.

File: src/task-actions.ts (hunk 2 above)
Comment (Nathan-SL) [reply src/task-actions.ts]: So, my fix was actually wrong. The issue only happens for a dockerized environment when we attempt to mount a non-absolute Cairo path. https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/307/commits/544ef8035f11adeecd134321baf27bfd08e6e668

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 4]
```diff
@@ -204,11 +204,20 @@ extendEnvironment((hre) => {
 
         const image = { repository, tag };
         const accountPaths = extractAccountPaths(hre);
+        const cairoPaths = [];
+        for (const cairoPath of hre.config.paths.cairoPaths || []) {
+            if (!path.isAbsolute(cairoPath)) {
+                cairoPaths.push(path.join(hre.config.paths.root, cairoPath));
```
[the remark this answers is not in the document]
Comment (Nathan-SL) [reply src/index.ts]: https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/307/commits/e371e3de5ba21c48b2978b93c9fe9a0510c41eed

File: test/configuration-tests/with-sources-path/hardhat.config.ts [hunk 5]
```diff
@@ -5,7 +5,8 @@ module.exports = {
         network: process.env.NETWORK
     },
     paths: {
-        starknetSources: "my-starknet-sources"
+        starknetSources: "my-starknet-sources",
+        cairoPaths: ["./new-sources"]
```
[the remark this answers is not in the document]
Comment (Nathan-SL) [reply test/configuration-tests/with-sources-path/hardhat.config.ts]: https://github.com/Shard-Labs/starknet-hardhat-plugin/pull/307/commits/9894adf26e4b0e812d296d7d39822b5156230a42

**Pull request merged**